### PR TITLE
BUG: make parser.pyx method polymorphic for gil/nogil cases

### LINKED ISSUE #27372: Cython warning about unsafe nogil cast in parsers.pyx
```
commit 681e6a9b07271a0955e6780e476ab2d7101e549c
Author: Phil Ruffwind <rf@rufflewind.com> 

    BUG: Segfault due to float_precision='round_trip'
    
    `round_trip` calls back into Python, so the GIL must be held.  It also
    fails to silence the Python exception, leading to spurious errors.
    Closes #15140. 
```    

681e6a9b07271a0955e6 closed #15140 but introduced a cython warning (seen with 0.29.12)
```
warning: pandas/_libs/parsers.pyx:1724:34: Casting a GIL-requiring 
function into a nogil function circumvents GIL validation
```

The commit says it fixes a segfault, while the cython warning suggests this is still unsafe.

The lines in question are https://github.com/pandas-dev/pandas/blob/8913b7e84ac75b1093bf708e0d1a9a0789e8cfcd/pandas/_libs/parsers.pyx#L1723-L1727

### PR DESCRIPTION
**Pull request by ghost**
- [x] closes #27372 
- [ ] Cython 0.30 released

**Warning**: this requires Cython 0.30 (not yet released as of Jul 2019) (the CI currently fails to build)

The [next release](https://cython.readthedocs.io/en/latest/src/changes.html#features-added) of Cython 0.30.0 [introduced](https://github.com/cython/cython/issues/2579) conditional gil acqusition with support for fused types.

This makes a clean solution for #27372 possible, eliminating the cython warning
```
warning: pandas/_libs/parsers.pyx:1724:34: Casting a GIL-requiring 
function into a nogil function circumvents GIL validation
```

### CODE REVIEW

File: pandas/_libs/parsers.pyx [hunk 1]
```diff
@@ -1692,6 +1692,25 @@ cdef:
...
+
+ctypedef float64_t (*double_converter_with_nogil_t)(
+              const char *, char **, char,
+              char, char, int, int *, int *) nogil;
+
+ctypedef fused double_converter_t:
+    double_converter_with_gil_t
+    double_converter_with_nogil_t
+
 cdef _try_double(parser_t *parser, int64_t col,
```
Comment (jbrockmendel) [inline pandas/_libs/parsers.pyx]: most of our linting tools won't work on these files, so we have to do it manually.  so: extra newline above this cdef, no semicolons above, pls align indentation on the double_converer_with definitions

File: pandas/_libs/parsers.pyx [hunk 2]
```diff
@@ -1692,6 +1692,25 @@ cdef:
     char* cneginf = b'-inf'
 
 
+ctypedef char flag_with_gil_t
+ctypedef float flag_no_gil_t
+
+ctypedef fused use_gil_t:
+    flag_with_gil_t
+    flag_no_gil_t
```
Comment (jbrockmendel) [inline pandas/_libs/parsers.pyx]: i dont understand what these are for, none of the new code above here is used again AFAICT.

File: pandas/_libs/parsers.pyx [hunk 3]
```diff
@@ -1752,18 +1766,40 @@ cdef inline int _try_double_nogil(parser_t *parser,
         char *p_end
         khiter_t k, k64
 
-    na_count[0] = 0
-    coliter_setup(&it, parser, col, line_start)
+    with nogil(double_converter_t is double_converter_with_nogil_t):
```
Comment (jbrockmendel) [inline pandas/_libs/parsers.pyx]: the motivating use case for conditional nogil was `with nogil(something_t is not object)`; is that pattern not available here?  (this file is one im less familiar with)

File: pandas/_libs/parsers.pyx (hunk 2 above)
Comment (pilkibun) [reply pandas/_libs/parsers.pyx]: left over from earlier approach which failed. removed.

File: pandas/_libs/parsers.pyx (hunk 3 above)
Comment (pilkibun) [reply pandas/_libs/parsers.pyx]: I'm not sure I understand your question. 

- Conditional gil is resolved at compile-time. 
- The use of the double_converter_t fused type generates two versions of this function. 
- This new `nogil(condition)` idiom is used to take the gil or not based on the specific type variant. 
- The condition is true for one generated version of the function and false for the other.

**Review by jreback — CHANGES_REQUESTED** [review]
this is highly unlikely to be looked at until cython is out for a while

**Comment by pilkibun** [comment]
> this is highly unlikely to be looked at until cython is out for a while

Of course.

**Comment by pilkibun** [comment]
@jbrockmendel, I'm clearing my desk of pandas PRs. This looks like it would be useful in the future, would you like to grab a local copy and take over or should I just delete?

**Comment by jbrockmendel** [comment]
@pilkibun is leaving this on the back-burner until cython 30 is out (plus some time for possible bugfix release)?

**Comment by pilkibun** [comment]
@jbrockmendel moving on to other interests and would like to to leave a clean slate. Best if you pull the branch and push it at your convenience sometime in the future. If you're not interested that's fine too of course.

**Comment by pilkibun** [comment]
Ok, someone will take care of this eventually.

**Pull request closed without merging**